contrib/TestHarness2/test_harness/summarize.py

```diff
@@ -0,0 +1,581 @@
+
+    def _call(self, callback: ParserCallback, attrs: Dict[str, str]) -> str | None:
+        try:
+            return callback(attrs)
+        except Exception as e:
+            _, _, exc_traceback = sys.exc_info()
+            child = SummaryTree('NonFatalParseError')
+            child.attributes['Severity'] = '30'
+            child.attributes['ErrorMessage'] = str(e)
+            child.attributes['Trace'] = repr(traceback.format_tb(exc_traceback))
+            self.out.append(child)
+            return None
+
+    def handle(self, attrs: Dict[str, str]):
+        if None in self.events:
+            for callback in self.events[None]:
+                self._call(callback, attrs)
+        for k, v in attrs.items():
+            if (k, None) in self.events:
+                for callback in self.events[(k, None)]:
+                    remap = self._call(callback, attrs)
+                    if remap is not None:
+                        v = remap
+                        attrs[k] = v
+            if (k, v) in self.events:
+                for callback in self.events[(k, v)]:
+                    remap = self._call(callback, attrs)
+                    if remap is not None:
+                        v = remap
+                        attrs[k] = v
+
+
+class Parser:
+    def parse(self, file: TextIO, handler: ParseHandler) -> None:
+        pass
+
+
+class XmlParser(Parser, xml.sax.handler.ContentHandler):
+    def __init__(self):
+        super().__init__()
+        self.handler: ParseHandler | None = None
+
+    def parse(self, file: TextIO, handler: ParseHandler) -> None:
+        xml.sax.parse(file, self)
+
+    def startElement(self, name, attrs) -> None:
+        attributes: Dict[str, str] = {}
+        for name in attrs.getNames():
+            attributes[name] = attrs.getValue(name)
+        assert self.handler is not None
+        self.handler.handle(attributes)
+
+
+class JsonParser(Parser):
+    def __init__(self):
+        super().__init__()
+
+    def parse(self, file: TextIO, handler: ParseHandler):
+        for line in file:
+            obj = json.loads(line)
+            handler.handle(obj)
+
+
+def format_test_error(attrs: Dict[str, str], include_details: bool) -> str:
+    trace_type = attrs['Type']
+    res = trace_type
+    if trace_type == 'InternalError':
+        res = '{} {} {}'.format(trace_type, attrs['File'], attrs['Line'])
+    elif trace_type == 'TestFailure':
+        res = '{} {}'.format(trace_type, attrs['Reason'])
+    elif trace_type == 'ValgrindError':
+        res = '{} {}'.format(trace_type, attrs['What'])
+    elif trace_type == 'ExitCode':
+        res = '{0} 0x{1:x}'.format(trace_type, int(attrs['Code']))
+    elif trace_type == 'StdErrOutput':
+        res = '{}: {}'.format(trace_type, attrs['Output'])
+    elif trace_type == 'BTreeIntegrityCheck':
+        res = '{}: {}'.format(trace_type, attrs['ErrorDetail'])
+    for k in ['Error', 'WinErrorCode', 'LinuxErrorCode']:
```

Comment:
Just in case you didn't see it, I recently added a new event that gets parsed by test harness: `RunningUnitTest`. It was added to make it easier to see what unit tests were run in a workload.

Reply:
I missed that, thanks for pointing it out